### PR TITLE
feat(backend/frontend/shared): Add notes to tasks bb-419

### PR DESCRIPTION
**Pull request by JD-Rosales**
### Summary

- Added check-note-access-to-task hook on backend.
- Added UI for viewing and adding notes.
- Added redux action for getting task notes and adding task notes.

https://github.com/user-attachments/assets/6c9f20e6-5c19-4cdf-8919-40d9b709c626

### CODE REVIEW

**Review by luiqor — COMMENTED** [review]
The branch has conflicts that need to be resolved, likely stemming from the "task modal actions".
After resolving these conflicts, please take a look at that: the Task-card component is also used in the modal. You'll need to disable notes button and change the icon color in the modal.

File: apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { type Knex } from "knex";
```
Comment (fshabanov) [inline apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts:1]: We already have a migration to add this table in main, it was added in a backend part of your task

File: apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+const TableName = {
+	TASK_NOTES: "task_notes",
```
Comment (marharita08) [inline apps/backend/src/db/migrations/20240918040515_add_task_notes_table.ts:4]: Did you add this table to db schema in readme file?

File: apps/backend/src/modules/tasks/libs/hooks/check-note-access-to-task.hook.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+const checkNoteAccessToTask =
+	(taskService: TaskService): APIPreHandler =>
+	async (request) => {
+		const { body, user } = request;
+
+		const { taskId } = body as TaskNoteRequestDto;
+
+		const task = await taskService.find(taskId);
+
+		if ((user as UserDto).id !== task?.userId) {
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/libs/hooks/check-note-access-to-task.hook.ts:19]: `checkAccessToTask` does exactly the same, remove this one

File: apps/backend/src/modules/tasks/task.controller.ts [hunk 4]
```diff
@@ -178,6 +267,49 @@ class TaskController extends BaseController {
 		};
 	}
 
+	/**
+	 * @swagger
+	 * /tasks/notes/:id:
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.controller.ts:272]:
```suggestion
	 * /tasks/:id/notes:
```

File: apps/backend/src/modules/tasks/libs/hooks/check-note-access-to-task.hook.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { ErrorMessage } from "~/libs/enums/enums.js";
+import { type APIPreHandler } from "~/libs/modules/controller/controller.js";
+import { HTTPCode } from "~/libs/modules/http/http.js";
+import { type UserDto } from "~/modules/users/users.js";
+
+import { type TaskService } from "../../task.service.js";
+import { TaskError } from "../exceptions/exceptions.js";
+import { type TaskNoteRequestDto } from "../types/types.js";
+
+const checkNoteAccessToTask =
```
Comment (marharita08) [inline apps/backend/src/modules/tasks/libs/hooks/check-note-access-to-task.hook.ts:10]: What difference between this and checkAccessToTask?

File: apps/backend/src/modules/tasks/task.repository.ts [hunk 6]
```diff
@@ -138,6 +163,23 @@ class TaskRepository implements Repository {
 		});
 	}
 
+	public async getNotesByTaskId(taskId: number): Promise<TaskNoteEntity[]> {
+		const notes = await this.taskNoteModel
+			.query()
+			.where({ taskId })
+			.orderBy("createdAt", "asc");
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.repository.ts:170]:
```suggestion
			.orderBy("createdAt", SortOrder.ASC);
```

File: apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useRef } from "~/libs/hooks/hooks.js";
+
+const handleClickOutside = <T extends HTMLElement>(
+	callback: () => void,
+): React.MutableRefObject<null | T> => {
+	const reference = useRef<null | T>(null);
+
+	useEffect(() => {
+		const handleClickOutside = (event: MouseEvent | TouchEvent): void => {
```
Comment (fshabanov) [inline apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts:9]: `handleClickOutside` is already defined outside, let's not have the same name

File: apps/frontend/src/modules/tasks/slices/tasks.slice.ts [hunk 8]
```diff
@@ -2,17 +2,25 @@ import { createSlice } from "@reduxjs/toolkit";
...
+	addNote,
+	getCurrentTasks,
+	getPastTasks,
+	getTaskNotes,
+	update,
+} from "./actions.js";
 
 type State = {
 	dataStatus: ValueOf<typeof DataStatus>;
+	task_notes: TaskNoteDto[];
```
Comment (fshabanov) [inline apps/frontend/src/modules/tasks/slices/tasks.slice.ts:17]:
```suggestion
	taskNotes: TaskNoteDto[];
```

File: apps/frontend/src/pages/tasks/libs/components/notes/notes.tsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+	onNoteClose,
+	onSubmit,
+	task,
+}: Properties) => {
+	const reference = handleClickOutside<HTMLDivElement>(onNoteClose);
+	const { control, errors, handleSubmit } = useAppForm<TaskNoteRequestDto>({
+		defaultValues: {
+			content: "",
+			taskId: task.id,
+		},
```
Comment (fshabanov) [inline apps/frontend/src/pages/tasks/libs/components/notes/notes.tsx:28]: Should be in a separate constant file

File: apps/frontend/src/pages/tasks/libs/components/task-card/task-card.tsx [hunk 10]
```diff
@@ -68,6 +97,14 @@ const TaskCard: React.FC<Properties> = ({
...
 				</div>
 			</div>
+
+			{isNoteOpen && (
+				<Notes
+					onNoteClose={handleNoteClose}
+					onSubmit={onAddTaskNote}
+					task={task}
+				/>
+			)}
```
Comment (fshabanov) [inline apps/frontend/src/pages/tasks/libs/components/task-card/task-card.tsx:107]: Notes shouldn't close when adding a new note

File: packages/shared/src/modules/tasks/libs/enums/tasks-api-path.enum.ts [hunk 11]
```diff
@@ -1,6 +1,8 @@
 const TasksApiPath = {
 	$ID: "/:id",
 	CURRENT: "/current",
+	NOTES: "/notes",
+	NOTES_$ID: "/notes/:id",
```
Comment (fshabanov) [inline packages/shared/src/modules/tasks/libs/enums/tasks-api-path.enum.ts:5]:
```suggestion
	NOTES_$ID: "/:id/notes",
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/icon/libs/maps/icon-name-to-svg.map.ts [hunk 12]
```diff
@@ -26,6 +28,7 @@ const iconNameToSvg: Record<
 	IconName,
 	React.FC<React.SVGProps<SVGSVGElement>>
 > = {
+	arrow_next: ArrowNext,
```
Comment (fshabanov) [inline apps/frontend/src/libs/components/icon/libs/maps/icon-name-to-svg.map.ts:31]:
```suggestion
	arrowNext: ArrowNext,
```
Keep the same naming styles

File: apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useRef } from "~/libs/hooks/hooks.js";
+
+const handleClickOutside = <T extends HTMLElement>(
+	callback: () => void,
+): React.MutableRefObject<null | T> => {
+	const reference = useRef<null | T>(null);
+
+	useEffect(() => {
+		const detectClickOutside = (event: MouseEvent | TouchEvent): void => {
```
Comment (fshabanov) [inline apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts:9]:
```suggestion
		const handleDetectClickOutside = (event: MouseEvent | TouchEvent): void => {
```

File: apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useRef } from "~/libs/hooks/hooks.js";
+
+const handleClickOutside = <T extends HTMLElement>(
+	callback: () => void,
```
Comment (fshabanov) [inline apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts:4]:
```suggestion
	onClickOutside: () => void,
```

File: apps/frontend/src/pages/tasks/libs/components/notes/notes.tsx [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+		return {
+			taskNotes: tasks.taskNotes,
+		};
+	});
+
+	const handleFormSubmit = useCallback(
+		(event: React.BaseSyntheticEvent): void => {
+			void handleSubmit(({ content, taskId }) => {
+				onSubmit({ content, taskId });
+			})(event);
```
Comment (fshabanov) [inline apps/frontend/src/pages/tasks/libs/components/notes/notes.tsx:42]:
```suggestion
			void handleSubmit(onSubmit)(event);
```

File: apps/frontend/src/pages/tasks/libs/components/notes/notes.tsx [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+	onSubmit: (payload: TaskNoteRequestDto) => void;
+	task: TaskDto;
+};
+
+const Notes: React.FC<Properties> = ({
+	onNoteClose,
+	onSubmit,
+	task,
+}: Properties) => {
+	const reference = handleClickOutside<HTMLDivElement>(onNoteClose);
```
Comment (fshabanov) [inline apps/frontend/src/pages/tasks/libs/components/notes/notes.tsx:24]: Reference for what? Give a better name

File: apps/frontend/src/pages/tasks/libs/components/task-card/task-card.tsx [hunk 17]
```diff
@@ -31,6 +48,14 @@ const TaskCard: React.FC<Properties> = ({
...
 	}, [task, onComplete]);
 
+	useEffect(() => {
+		onGetTaskNotes(task.id);
+	}, [onGetTaskNotes, task.id]);
+
+	useEffect(() => {
+		onGetTaskNotes(task.id);
+	}, [onGetTaskNotes, task.id]);
+
```
Comment (fshabanov) [inline apps/frontend/src/pages/tasks/libs/components/task-card/task-card.tsx:58]: Duplicate

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useRef } from "~/libs/hooks/hooks.js";
+
+const handleClickOutside = <T extends HTMLElement>(
+	onClickOutside: () => void,
+): React.MutableRefObject<null | T> => {
+	const reference = useRef<null | T>(null);
```
Comment (fshabanov) [inline apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts:6]: Let's send a reference as an argument

File: apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useRef } from "~/libs/hooks/hooks.js";
+
+const handleClickOutside = <T extends HTMLElement>(
```
Comment (fshabanov) [inline apps/frontend/src/libs/helpers/handle-click-outside/handle-click-outside.helper.ts:3]:
```suggestion
const useClickOutside = <T extends HTMLElement>(
```
Let's make it a hook

File: apps/frontend/src/pages/tasks/libs/components/task-card/task-card.tsx [hunk 20]
```diff
@@ -70,6 +91,23 @@ const TaskCard: React.FC<Properties> = ({
...
+									hasVisuallyHiddenLabel
+									iconName="note"
+									iconPosition="left"
+									label="notes"
+									onClick={handleNoteOpen}
+									variant="icon"
+								/>
+								<Button
+									hasVisuallyHiddenLabel
+									iconColor=""
```
Comment (fshabanov) [inline apps/frontend/src/pages/tasks/libs/components/task-card/task-card.tsx:104]:
```suggestion
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Comment by JD-Rosales** [comment]
Submit button is currently not visible because of a bug in icon button variant. @KeidsID is working a fix on that.

**Review by fshabanov — APPROVED** [review]

**Pull request merged**